src/libraries/System.Collections/src/System/Collections/Generic/ReadOnlySet.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Collections.ObjectModel
+{
+    /// <summary>Represents a read-only, generic set of values.</summary>
+    /// <typeparam name="T">The type of values in the set.</typeparam>
+    [DebuggerDisplay("Count = {Count}")]
+    public class ReadOnlySet<T> : IReadOnlySet<T>, ISet<T>, ICollection
+    {
+        /// <summary>The wrapped set.</summary>
+        private readonly ISet<T> _set;
+
+        /// <summary>Initializes a new instance of the <see cref="ReadOnlySet{T}"/> class that is a wrapper around the specified set.</summary>
+        /// <param name="set">The set to wrap.</param>
+        public ReadOnlySet(ISet<T> set)
+        {
+            ArgumentNullException.ThrowIfNull(set);
+            _set = set;
+        }
+
+        /// <summary>Gets an empty <see cref="ReadOnlySet{T}"/>.</summary>
+        public static ReadOnlySet<T> Empty { get; } = new ReadOnlySet<T>(new HashSet<T>());
```

Comment:
Should this be able to use `HashSet<T>.Empty`?

Reply:
Seemed there's no `Empty` for `HashSet`
https://apisof.net/catalog/dc5e262d8b718b155a12f079686ad57c

Reply:
indeed, it was a question for @stephentoub - is that intentionally missing, or should we add it? I'm assuming it's an intentional decision to not have it on any mutable collections, although we do have it in some places eg., XmlDictionary.

Reply:
We intentionally do not have it on mutable collections. Empty would return a singleton `HashSet<T>`, which could then be mutated and no longer be empty.

(XmlDictionary.Empty doesn't return XmlDictionary, it returns IXmlDictionary.)